### ISSUE TITLE
S3 backend with ZIP download broken?

### ISSUE THREAD

**Issue by bgst** [filesender/filesender#1054]
It seems that streaming from an S3 backend does not work, i.e. downloading a ZIP file does not work when the files are stored on S3. Single file download works without issues. I have changed the code to use single buckets for storage instead of creating a single bucket for each file, but the original code has the same issue as the modified one. Tried both with Wasabi and Backblaze. A guess is that this relates to the provided streamwrapper.

Is this a known issue?

**Comment by bgst** [comment]
The issue relates to the \storage\StorageCloudS3::readChunk() method ignoring the passed $length param and thus typically delivering a complete object via $client->getObject(), which then in turn is stripped down to e.g. 8192 bytes by fread() in the ZipStreamer->streamFileData() method truncating the returned data.

One quick fix for this is simply to ensure that the readChunk() method returns the count amount of bytes requested. This will be very ineffective when using a default 8192 bytes from a remote location with S3. Better is probably to replace the current implementation of FileSender's StorageCloudS3 with Amazon's StreamWrapper implementation from their AWS SDK.

**Comment by monkeyiq** [comment]
I think the streaming of unencrypted files in zip would be going through StorageCloudS3Stream with it's `stream_read()` calling `StorageCloudS3::readChunk()`. If the streaming implementation that is calling to `stream_read()` is only using 8k blocks that is unfortunate. It would be possible though for `stream_read()` to hold a cache for a larger read and hand that back to the caller 8k at a time.

**Comment by bgst** [comment]
Exactly, very ineffective. I implemented a file based cache system and a separate task kicked in by cron that wacks any temporarily cached files (on file system) that are older than $config['storage_cache_max_age'] time, which I have defaulted to six hours. Not that cool version, but it is one way to reduce roundtrips to S3. Risks are of course that (1) filesystem grows too large too fast and there is not enough space for the cache and (2) very slow downloads will find their chunks disappearing (probability eventually low) under their feet.

**Comment by kerins** [comment]
We have run into this issue and are currently looking at implementing a work around. Thanks for the inputs

**Comment by kerins** [comment]
By setting 
stream_set_chunk_size($stream, self::STREAM_CHUNK_SIZE);
in lib/PHPZipStreamer/src/ZipStreamer.php
on the line before 403 
while (!feof($stream) && $data = fread($stream, self::STREAM_CHUNK_SIZE)) {

I got this to work.

**Comment by monkeyiq** [comment]
@kerins does https://github.com/filesender/filesender/pull/1091 do a similar thing for you?

The ZipStreamer and FileSender chunk sizes are different so another improvement might be to set them to the same. This could be tested by modifing the STREAM_CHUNK_SIZE constant in ZipStreamer.php to be the same as the FileSender chunk size (upload_chunk_size from config.php). Since the zip is only to store rather than deflate the fixed size of STREAM_CHUNK_SIZE in the zip class doesn't really apply and might as well just be the filesender chunk size.

**Comment by monkeyiq** [comment]
I also made https://github.com/filesender/filesender/pull/1093 to test if making STREAM_CHUNK_SIZE match the filesender chunk size offers any more improvements. I might try to push the update from 1093 upstream as it may help other network backed filesender storage cases too.

**Comment by kerins** [comment]
Super work. These 2 changed did indeed fix the zip changes. After testing my fix with files larger than the chunk size I was having issues. in S3 the files in 5.2 MB chunk sizes and files larger than this where causing really weird issues with the Zip stream, but your fix works like a charm. 

Tar function seems to suffer from a similar issue.

I also pulled the latest version of the 2021 bootstrap with these fixes merged and it appears broken, the first issue was i needed php7.4-gmp installed

and then I got the following error :

[php7:notice] [pid 347257] PHP Notice:  Trying to access array offset on value of type null in /opt/filesender/optional-dependencies/s3/vendor/aws/aws-sdk-php/src/RetryMiddleware.php on line 60, referer: 
 [php7:notice] [pid 347257]  PHP Notice:  Trying to access array offset on value of type null in /opt/filesender/optional-dependencies/s3/vendor/aws/aws-sdk-php/src/RetryMiddleware.php on line 60, referer: 
[php7:warn] [pid 347257] PHP Warning:  Cannot modify header information - headers already sent by (output started at /opt/filesender/lib/vendor/barracudanetworks/archivestream-php/src/Archive.php:355) in /opt/filesender/www/download.php on line 136, referer: 

I gave up at this point and as a quick and dirty solution I copied the 
lib/PHPZipStreamer directory back into lib
and replaced classes/utils/Archiver.class.php with my older version and it works again for me.

thanks again for your help with this. It is currently in a useable state for us
